src/provider/mlkzg.rs

```diff
@@ -0,0 +1,621 @@
+      // state as that of the verifier
+      let _d_0 = Self::verifier_second_challenge(&C_B, &w, transcript);
+
+      (w, v)
+    };
+
+    ///// END helper closures //////////
+
+    let ell = x.len();
+    let n = hat_P.len();
+    assert_eq!(n, 1 << ell); // Below we assume that n is a power of two
+
+    // Phase 1  -- create commitments com_1, ..., com_\ell
+    let mut polys: Vec<Vec<E::Fr>> = Vec::new();
+    polys.push(hat_P.to_vec());
+    for i in 0..ell {
+      let Pi_len = polys[i].len() / 2;
+      let mut Pi = vec![E::Fr::ZERO; Pi_len];
+
+      #[allow(clippy::needless_range_loop)]
+      for j in 0..Pi_len {
+        Pi[j] = x[ell-i-1] * polys[i][2*j + 1]            // Odd part of P^(i-1)
+                      + (E::Fr::ONE - x[ell-i-1]) * polys[i][2*j]; // Even part of P^(i-1)
+      }
+
+      if i == ell - 1 && *eval != Pi[0] {
+        return Err(NovaError::UnSat);
+      }
+
+      polys.push(Pi);
+    }
+
+    // We do not need to commit to the first polynomial as it is already committed.
+    // Compute commitments in parallel
+    let com: Vec<E::G1Affine> = (1..polys.len())
+      .into_par_iter()
+      .map(|i| {
+        <NE::CE as CommitmentEngineTrait<NE>>::commit(ck, &polys[i])
+          .comm
+          .preprocessed()
+      })
+      .collect();
+
+    // Phase 2
+    // We do not need to add x to the transcript, because in our context x was
+    // obtained from the transcript.
+    let r = Self::compute_challenge(&C.comm.preprocessed(), eval, &com, transcript);
+    let u = vec![r, -r, r * r];
+
+    // Phase 3 -- create response
+    let mut com_all = com.clone();
+    com_all.insert(0, C.comm.preprocessed());
+    let (w, v) = kzg_open_batch(&com_all, &polys, &u, transcript);
+
+    Ok(EvaluationArgument { com, w, v })
+  }
+
+  /// A method to verify purported evaluations of a batch of polynomials
+  fn verify(
+    vk: &Self::VerifierKey,
+    transcript: &mut <NE as NovaEngine>::TE,
+    C: &Commitment<NE>,
+    point: &[E::Fr],
+    P_of_x: &E::Fr,
+    pi: &Self::EvaluationArgument,
+  ) -> Result<(), NovaError> {
+    let x = point.to_vec();
+    let y = P_of_x;
+
+    // vk is hashed in transcript already, so we do not add it here
+
+    let kzg_verify_batch = |vk: &KZGVerifierKey<E>,
+                            C: &Vec<E::G1Affine>,
+                            W: &Vec<E::G1Affine>,
+                            u: &Vec<E::Fr>,
+                            v: &Vec<Vec<E::Fr>>,
+                            transcript: &mut <NE as NovaEngine>::TE|
+     -> bool {
+      let k = C.len();
+      let t = u.len();
+
+      let q = Self::get_batch_challenge(C, u, v, transcript);
+      let q_powers = Self::batch_challenge_powers(q, k); // 1, q, q^2, ..., q^(k-1)
+
+      // Compute the commitment to the batched polynomial B(X)
+      let c_0: E::G1 = C[0].into();
+      let C_B = (c_0 + NE::GE::vartime_multiscalar_mul(&q_powers[1..k], &C[1..k])).preprocessed();
```

Comment:
This is not required since it's only used as an intermediary computation for `L` later on

Reply:
Fixed in https://github.com/lurk-lab/arecibo/pull/231